### PR TITLE
planner: add more test cases for outer join's leading hint

### PR DESCRIPTION
**Pull request by Reminiscent**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/29932

Problem Summary:
add more test cases for outer join's leading hint

### What is changed and how it works?
add more test cases for outer join's leading hint

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/testdata/join_reorder_suite_out.json [hunk 1]
```diff
@@ -2516,137 +2516,1924 @@
...
+          "    │   └─TableFullScan 10000.00 cop[tikv] table:t2, partition:p2 keep order:false, stats:pseudo",
+          "    ├─TableReader 9990.00 root  data:Selection",
+          "    │ └─Selection 9990.00 cop[tikv]  not(isnull(test.t2.b))",
+          "    │   └─TableFullScan 10000.00 cop[tikv] table:t2, partition:p3 keep order:false, stats:pseudo",
+          "    └─TableReader 9990.00 root  data:Selection",
+          "      └─Selection 9990.00 cop[tikv]  not(isnull(test.t2.b))",
+          "        └─TableFullScan 10000.00 cop[tikv] table:t2, partition:p4 keep order:false, stats:pseudo"
+        ],
+        "Warning": [
+          "Warning 1815 leading hint is inapplicable, check if the leading hint table is valid"
```
Comment (Reminiscent) [inline planner/core/testdata/join_reorder_suite_out.json:2637]: Should document this part on the user doc.

File: planner/core/testdata/join_reorder_suite_out.json [hunk 2]
```diff
@@ -2516,431 +2516,4233 @@
...
```
Comment (Reminiscent) [inline planner/core/testdata/join_reorder_suite_out.json:2817]: Check here.

File: planner/core/testdata/join_reorder_suite_out.json [hunk 3]
```diff
@@ -2603,339 +3942,512 @@
...
-      }
-    ]
-  },
-  {
-    "Name": "TestJoinOrderHint4DifferentJoinType",
-    "Cases": [
+      },
       {
-        "SQL": "select /*+ straight_join() */ * from t1 join t2 on t1.a=t2.a join t3 on t2.b=t3.b",
+        "SQL": "select /*+ leading(t1, t2) */ * from t4 join t on t4.a=t.a left join t1 on t.a = t1.a join t2 on t1.b = t2.b join t3 on t2.b=t3.b;",
```
Comment (Reminiscent) [inline planner/core/testdata/join_reorder_suite_out.json:3947]: Check here.

**Review by qw4990 — APPROVED** [review]

**Review by xuyifangreeneyes — APPROVED** [review]
rest LGTM

File: planner/core/testdata/join_reorder_suite_in.json [hunk 4]
```diff
@@ -302,62 +401,117 @@
...
+      "select /*+ leading(t3, t2, t) */ * from t4 join t on t4.a=t.a left join t1 on t.a = t1.a join t2 on t.b = t2.b join t3 on t2.b=t3.b;",
+
+      "select /*+ leading(t3) */ * from t2 left join (t1 left join t3 on t1.a=t3.a) on t2.b=t1.b;",
+      "select /*+ leading(t2, t1, t3) */ * from t2 left join (t1 left join t3 on t1.a=t3.a) on t2.b=t1.b;",
+
+      "select /*+ leading(t2, t3) */ * from t2 left join (t1 join t3 on t1.a=t3.a join t4 on t3.b = t4.b) on t2.b=t1.b;",
+      "select /*+ leading(t3, t4) */ * from t2 left join (t1 join t3 on t1.a=t3.a join t4 on t3.b = t4.b) on t2.b=t1.b;",
+
+      "select /*+ leading(t3, t4) */ * from t2 left join (t1 join t3 on t1.a=t3.a join t4 on t3.b = t4.b) on t2.b=t1.b join t5 on t2.a = t5.a join t6 on t5.b=t6.b;",
+      "select /*+ leading(t3, t4) leading(t5, t6) */ * from t2 left join (t1 join t3 on t1.a=t3.a join t4 on t3.b = t4.b) on t2.b=t1.b join t5 on t2.a = t5.a join t6 on t5.b=t6.b;", // should support this case ;ater
```
[the remark this answers is not in the document]
Comment (Reminiscent) [reply planner/core/testdata/join_reorder_suite_in.json:419]: Yes.

**Pull request merged**